### PR TITLE
[SPARK-13295] [ ML, MLlib ] AFTSurvivalRegression.AFTAggregator improvements - avoid creating new instances of arrays/vectors for each record

### PR DESCRIPTION
**Pull request by NarineK**
As also mentioned/marked by TODO in AFTAggregator.AFTAggregator.add(data: AFTPoint) method a new array is being created for intercept value and it is being concatenated
with another array which contains the betas, the resulted Array is being converted into a Dense vector which in its turn is being converted into breeze vector.
This is expensive and not necessarily beautiful.

I've tried to solve above mentioned problem by simple algebraic decompositions - keeping and treating intercept independently.

Please let me know what do you think and if you have any questions.

Thanks,
Narine

### CODE REVIEW

**Comment by mengxr** [comment]
@yanboliang Could you take a look?

**Comment by mengxr** [comment]
ok to test

File: mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala [hunk 1]
```diff
@@ -438,13 +438,15 @@ private class AFTAggregator(parameters: BDV[Double], fitIntercept: Boolean)
   extends Serializable {
 
   // beta is the intercept and regression coefficients to the covariates
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala]: Please update the doc, beta is only the regression coefficients after your change.

File: mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala [hunk 2]
```diff
@@ -453,7 +455,8 @@ private class AFTAggregator(parameters: BDV[Double], fitIntercept: Boolean)
 
   // Here we optimize loss function over beta and log(sigma)
   def gradient: BDV[Double] = BDV.vertcat(BDV(Array(gradientLogSigmaSum / totalCnt.toDouble)),
-    gradientBetaSum/totalCnt.toDouble)
+    BDV.vertcat(BDV(Array(gradientInterceptSum/totalCnt.toDouble)),
+                gradientBetaSum/totalCnt.toDouble))
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala]: Use ```BDV.vertcat(BDV(Array(gradientLogSigmaSum / totalCnt.toDouble)), BDV(Array(gradientInterceptSum/totalCnt.toDouble)), gradientBetaSum/totalCnt.toDouble)``` to combine three parts directly.

File: mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala [hunk 3]
```diff
@@ -464,15 +467,12 @@ private class AFTAggregator(parameters: BDV[Double], fitIntercept: Boolean)
    */
   def add(data: AFTPoint): this.type = {
 
-    // TODO: Don't create a new xi vector each time.
-    val xi = if (fitIntercept) {
-      Vectors.dense(Array(1.0) ++ data.features.toArray).toBreeze
-    } else {
-      Vectors.dense(Array(0.0) ++ data.features.toArray).toBreeze
-    }
+    val fitInterceptFlag = if (fitIntercept) 1.0 else 0.0
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala]: ```fitInterceptFlag``` -> ```interceptFlag```. ```fitInterceptFlag``` is more like a name of boolean variable.

File: mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala [hunk 4]
```diff
@@ -481,8 +481,10 @@ private class AFTAggregator(parameters: BDV[Double], fitIntercept: Boolean)
     assert(!lossSum.isInfinity,
       s"AFTAggregator loss sum is infinity. Error for unknown reason.")
 
-    gradientBetaSum += xi * (delta - math.exp(epsilon)) / sigma
-    gradientLogSigmaSum += delta + (delta - math.exp(epsilon)) * epsilon
+    val delta_expeps = delta - math.exp(epsilon)
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala]: ```delta_expeps``` -> ```deltaMinusExpEps```, we use the camel case convention.

**Comment by yanboliang** [comment]
@NarineK Thanks for working on this issue, it looks good overall and I left some inline comments.

**Comment by NarineK** [comment]
Thank you for the review comments, @yanboliang 
I've added your suggestions. Let me know if you have more comments.

File: mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala [hunk 5]
```diff
@@ -437,23 +437,25 @@ object AFTSurvivalRegressionModel extends MLReadable[AFTSurvivalRegressionModel]
...
+  private var gradientCoefficientSum = BDV.zeros[Double](coefficients.length)
+  private var gradientInterceptSum = 0.0
   private var gradientLogSigmaSum = 0.0
 
   def count: Long = totalCnt
 
   def loss: Double = if (totalCnt == 0) 1.0 else lossSum / totalCnt
 
-  // Here we optimize loss function over beta and log(sigma)
+  // Here we optimize loss function over coefficients and log(sigma)
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala]: ```beta``` means ```coefficients and intercept```, so here should be ```coefficients, intercept and log(sigma)``` in the annotation.

**Comment by yanboliang** [comment]
LGTM except minor issue, thanks!

**Comment by NarineK** [comment]
Fixed the comment, thnx @yanboliang

File: mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala [hunk 6]
```diff
@@ -437,23 +437,25 @@ object AFTSurvivalRegressionModel extends MLReadable[AFTSurvivalRegressionModel]
...
 
   def count: Long = totalCnt
 
   def loss: Double = if (totalCnt == 0) 1.0 else lossSum / totalCnt
 
-  // Here we optimize loss function over beta and log(sigma)
+  // Here we optimize loss function over coefficients, intercept and log(sigma)
   def gradient: BDV[Double] = BDV.vertcat(BDV(Array(gradientLogSigmaSum / totalCnt.toDouble)),
-    gradientBetaSum/totalCnt.toDouble)
+    BDV(Array(gradientInterceptSum/totalCnt.toDouble)), gradientCoefficientSum/totalCnt.toDouble)
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala]: Vector ```vertcat``` may be time-consuming, we can fix this issue use pre-assigned array. But it also involves other changes which are beyond the scope of this PR. I will fix this when solving [SPARK-13322](https://issues.apache.org/jira/browse/SPARK-13322), this PR is ready to merge, thanks.

**Comment by yanboliang** [comment]
LGTM, thanks @NarineK 
ping @mengxr Could you take a look when you have time? [SPARK-13322](https://issues.apache.org/jira/browse/SPARK-13322) depends on this PR.

**Comment by mengxr** [comment]
Merged into master. Thanks!

**Pull request closed without merging**